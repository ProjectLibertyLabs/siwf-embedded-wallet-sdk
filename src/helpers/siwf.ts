import { SignatureFn } from "../types";
import { SiwfResponsePayload, SiwfResponse } from "@projectlibertylabs/siwf";
import { encodedValueToSignature, accountIdToPublicKey } from "./utils";

export interface CreateLoginSiwfResponseArguments
  extends Record<string, unknown> {
  domain: string;
  uri: string;
  version: string;
  chainId: string;
  nonce?: string;
  issuedAt?: string;
}

export async function createLoginSiwfResponse(
  accountId: string,
  signatureFn: SignatureFn,
  payloadArguments: CreateLoginSiwfResponseArguments,
): Promise<SiwfResponse> {
  const loginCaip122 = `${payloadArguments.domain} wants you to sign in with your Frequency account:
    frequency:${payloadArguments.chainId}:${accountId}
    
    URI: ${payloadArguments.uri}
    Version: ${payloadArguments.version}
    Nonce: ${payloadArguments.nonce}
    Chain ID: frequency:${payloadArguments.chainId}
    Issued At: ${payloadArguments.issuedAt}`;

  const encodedValue = await signatureFn({
    method: "personal_sign",
    params: [accountId, loginCaip122],
  });

  const userPublicKey = accountIdToPublicKey(accountId);

  const signature = encodedValueToSignature(encodedValue);

  return {
    userPublicKey,
    payloads: [
      {
        signature,
        type: "login",
        payload: {
          message: loginCaip122,
        },
      },
    ],
    credentials: [],
  };
}

export async function createSignInSiwfResponse(
  accountId: string,
  payloads: SiwfResponsePayload[],
): Promise<SiwfResponse> {
  const userPublicKey = accountIdToPublicKey(accountId);
  return {
    userPublicKey,
    payloads,
    credentials: [],
  };
}
