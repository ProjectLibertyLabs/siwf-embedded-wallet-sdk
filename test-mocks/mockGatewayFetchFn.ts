import { GatewayFetchFn } from "../src/types/param-types";
import {
  AccountResponse,
  ChainInfoResponse,
  GatewaySiwfResponse,
} from "../src/types/response-types";

export const mockGatewayFetchFactory = (
  hasAccountResponse: AccountResponse | null,
  providerAccountResponse: AccountResponse | null,
  gatewaySiwfResponse: GatewaySiwfResponse,
  chainInfoResponse: ChainInfoResponse,
  providerControlKey: string,
  finalResponse: { response: AccountResponse | null } = { response: null },
): GatewayFetchFn => {
  return async (method, path) => {
    if (path === "/v2/accounts/siwf") {
      return {
        ok: true,
        status: 200,
        json: async () => gatewaySiwfResponse,
      } as Response;
    }

    if (path.startsWith("/v1/accounts/account/")) {
      if (method === "GET")
        return {
          ok: true,
          status: 200,
          json: async () => {
            if (path.includes(providerControlKey))
              return providerAccountResponse;
            return finalResponse.response || hasAccountResponse;
          },
        } as Response;
    }

    if (path.startsWith("/v1/frequency/blockinfo")) {
      if (method === "GET")
        return {
          ok: true,
          status: 200,
          json: async () => chainInfoResponse,
        } as Response;
    }

    return new Response(`Error: ${path} not found`, {
      status: 404,
    });
  };
};
