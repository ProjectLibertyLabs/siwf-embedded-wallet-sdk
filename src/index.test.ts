import { describe, it, expect } from "vitest";
import { startSiwf } from "./index.js";
import { mockGatewayFetchFactory } from "./tests/mockGatewayFetchFn";
import {
  mockNewUserAccountResponse,
  mockNewUserGatewaySiwfResponse,
  mockProviderAccountResponse,
  mockProviderEncodedRequest,
  mockRetunringUserGatewaySiwfResponse,
  mockReturningUserAccountResponse,
  mockUserAddress,
} from "./tests/consts";
import { decodeSignedRequest } from "@projectlibertylabs/siwf";

const providerControlKey = decodeSignedRequest(mockProviderEncodedRequest)
  .requestedSignatures.publicKey.encodedValue;

describe("Basic startSiwf test", () => {
  it("Can login", async () => {
    const resp = await startSiwf(
      mockUserAddress,
      async () => "0xdef0",
      mockGatewayFetchFactory(
        mockReturningUserAccountResponse,
        mockProviderAccountResponse,
        mockRetunringUserGatewaySiwfResponse,
        providerControlKey,
      ),
      mockProviderEncodedRequest,
      "JohnDoe",
      "john.doe@example.com",
      () => {},
    );
    expect(resp.controlKey).toEqual(mockUserAddress);
    expect(resp.msaId).toEqual(mockRetunringUserGatewaySiwfResponse.msaId);
    expect(resp).toMatchSnapshot();
  });

  it("Throws if provider account not found", async () => {
    await expect(
      startSiwf(
        mockUserAddress,
        async () => "0xdef0",
        mockGatewayFetchFactory(
          mockReturningUserAccountResponse,
          null,
          mockRetunringUserGatewaySiwfResponse,
          providerControlKey,
        ),
        mockProviderEncodedRequest,
        "JohnDoe",
        "john.doe@example.com",
        () => {},
      ),
    ).rejects.toThrow("Unable to find provider account!");
  });

  it("If new user, throws if no signUpHandle", async () => {
    await expect(
      startSiwf(
        mockUserAddress,
        async () => "0xdef0",
        mockGatewayFetchFactory(
          mockNewUserAccountResponse,
          mockProviderAccountResponse,
          mockNewUserGatewaySiwfResponse,
          providerControlKey,
        ),
        mockProviderEncodedRequest,
        null,
        "john.doe@example.com",
        () => {},
      ),
    ).rejects.toThrow("signUpHandle missing for non-existent account.");
  });

  it("If new user, throws if no signUpEmail", async () => {
    await expect(
      startSiwf(
        mockUserAddress,
        async () => "0xdef0",
        mockGatewayFetchFactory(
          mockNewUserAccountResponse,
          mockProviderAccountResponse,
          mockNewUserGatewaySiwfResponse,
          providerControlKey,
        ),
        mockProviderEncodedRequest,
        "JohnDoe",
        null,
        () => {},
      ),
    ).rejects.toThrow("signUpEmail missing for non-existent account.");
  });

  it("Can sign up", async () => {
    const resp = await startSiwf(
      mockUserAddress,
      async () => "0xdef0",
      mockGatewayFetchFactory(
        mockNewUserAccountResponse,
        mockProviderAccountResponse,
        mockNewUserGatewaySiwfResponse,
        providerControlKey,
      ),
      mockProviderEncodedRequest,
      "JohnDoe",
      "john.doe@example.com",
    );

    expect(resp.controlKey).toEqual(mockUserAddress);
    expect(resp.msaId).toEqual(mockNewUserGatewaySiwfResponse.msaId);
    expect(resp).toMatchSnapshot();
  });
});
