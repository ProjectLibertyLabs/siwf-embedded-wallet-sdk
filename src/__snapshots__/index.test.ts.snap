// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Basic startSiwf test > Can do a thing 1`] = `
{
  "controlKey": "0xf24FF3a9CF04c71Dbc94D0b566f7A27B94566cac",
  "email": "user@example.com",
  "graphKey": {
    "encodedPrivateKeyValue": "0xd0910c853563723253c4ed105c08614fc8aaaf1b0871375520d72251496e8d87",
    "encodedPublicKeyValue": "0xb5032900293f1c9e5822fd9c120b253cb4a4dfe94c214e688e01f32db9eedf17",
    "encoding": "base16",
    "format": "bare",
    "id": "did:key:z6QNucQV4AF1XMQV4kngbmnBHwYa6mVswPEGrkFrUayhttT1",
    "keyType": "dsnp.public-key-key-agreement",
    "type": "X25519",
  },
  "phoneNumber": "555-867-5309",
  "rawCredentials": [
    {
      "@context": [
        "https://www.w3.org/ns/credentials/v2",
        "https://www.w3.org/ns/credentials/undefined-terms/v2",
      ],
      "credentialSchema": {
        "id": "https://schemas.frequencyaccess.com/VerifiedEmailAddressCredential/bciqe4qoczhftici4dzfvfbel7fo4h4sr5grco3oovwyk6y4ynf44tsi.json",
        "type": "JsonSchema",
      },
      "credentialSubject": {
        "emailAddress": "john.doe@example.com",
        "id": "did:key:z6QNucQV4AF1XMQV4kngbmnBHwYa6mVswPEGrkFrUayhttT1",
        "lastVerified": "2024-08-21T21:27:59.309+0000",
      },
      "issuer": "did:web:frequencyaccess.com",
      "proof": {
        "cryptosuite": "eddsa-rdfc-2022",
        "proofPurpose": "assertionMethod",
        "proofValue": "z4jArnPwuwYxLnbBirLanpkcyBpmQwmyn5f3PdTYnxhpy48qpgvHHav6warjizjvtLMg6j3FK3BqbR2nuyT2UTSWC",
        "type": "DataIntegrityProof",
        "verificationMethod": "did:web:frequencyaccess.com#z6MkofWExWkUvTZeXb9TmLta5mBT6Qtj58es5Fqg1L5BCWQD",
      },
      "type": [
        "VerifiedEmailAddressCredential",
        "VerifiableCredential",
      ],
      "validFrom": "2024-08-21T21:28:08.289+0000",
    },
    {
      "@context": [
        "https://www.w3.org/ns/credentials/v2",
        "https://www.w3.org/ns/credentials/undefined-terms/v2",
      ],
      "credentialSchema": {
        "id": "https://schemas.frequencyaccess.com/VerifiedGraphKeyCredential/bciqmdvmxd54zve5kifycgsdtoahs5ecf4hal2ts3eexkgocyc5oca2y.json",
        "type": "JsonSchema",
      },
      "credentialSubject": {
        "encodedPrivateKeyValue": "0xd0910c853563723253c4ed105c08614fc8aaaf1b0871375520d72251496e8d87",
        "encodedPublicKeyValue": "0xb5032900293f1c9e5822fd9c120b253cb4a4dfe94c214e688e01f32db9eedf17",
        "encoding": "base16",
        "format": "bare",
        "id": "did:key:z6QNucQV4AF1XMQV4kngbmnBHwYa6mVswPEGrkFrUayhttT1",
        "keyType": "dsnp.public-key-key-agreement",
        "type": "X25519",
      },
      "issuer": "did:key:z6QNucQV4AF1XMQV4kngbmnBHwYa6mVswPEGrkFrUayhttT1",
      "proof": {
        "cryptosuite": "eddsa-rdfc-2022",
        "proofPurpose": "assertionMethod",
        "proofValue": "z2HHWwtWggZfvGqNUk4S5AAbDGqZRFXjpMYAsXXmEksGxTk4DnnkN3upCiL1mhgwHNLkxY3s8YqNyYnmpuvUke7jF",
        "type": "DataIntegrityProof",
        "verificationMethod": "did:key:z6MktZ15TNtrJCW2gDLFjtjmxEdhCadNCaDizWABYfneMqhA",
      },
      "type": [
        "VerifiedGraphKeyCredential",
        "VerifiableCredential",
      ],
      "validFrom": "2024-08-21T21:28:08.289+0000",
    },
  ],
  "signUpReferenceId": "MjY3MjI3NWZlMGM0NTZmYjY3MWU0ZjQxN2ZiMmY5ODkyYzc1NzNiYQo",
  "signUpStatus": "waiting",
}
`;
